web/components/contract-search.tsx

```diff
@@ -103,26 +91,73 @@ export function ContractSearch(props: {
     cardHideOptions,
     highlightOptions,
     headerClassName,
-    useQuerySortLocalStorage,
+    persistPrefix,
     useQuerySortUrlParams,
   } = props
 
-  const [numPages, setNumPages] = useState(1)
-  const [pages, setPages] = useState<Contract[][]>([])
-  const [showTime, setShowTime] = useState<ShowTime | undefined>()
+  const store = safeSessionStorage()
+  const persistAs = (name: string) => {
+    return persistPrefix ? { prefix: persistPrefix, name, store } : undefined
+  }
+
+  const [numPages, setNumPages] = usePersistentState(1, persistAs('numPages'))
+  const [pages, setPages] = usePersistentState<Contract[][]>(
+    [],
+    persistAs('pages')
+  )
+  const [showTime, setShowTime] = usePersistentState<ShowTime | null>(
+    null,
+    persistAs('showTime')
+  )
 
-  const searchParameters = useRef<SearchParameters | undefined>()
+  const searchParameters = useRef<SearchParameters | null>(null)
   const requestId = useRef(0)
 
+  useLayoutEffect(() => {
+    if (persistPrefix && store) {
+      const parameters = loadState(getKey(persistPrefix, 'parameters'), store)
+      const scrollY = loadState(getKey(persistPrefix, 'scrollY'), store)
+      if (parameters !== undefined) {
+        console.log('Restoring search parameters: ', parameters)
+        searchParameters.current = parameters as SearchParameters
+      }
+      if (scrollY !== undefined) {
+        console.log('Restoring scroll position: ', scrollY)
+        window.scrollTo(0, scrollY as number)
+      }
+    }
+  }, [])
+
+  useEffect(() => {
+    if (persistPrefix && store) {
+      const handleScroll = (e: Event) => {
+        const scrollY = (e.currentTarget as Window).scrollY
+        console.log('Saving scroll position: ', scrollY)
+        saveState(getKey(persistPrefix, 'scrollY'), scrollY, store)
+      }
+      window.addEventListener('scroll', handleScroll)
```

Comment:
This kind of listener is pretty slow. Can't we measure the scroll on navigating away?

Reply:
I agree, it was the simplest thing but that would be a much better implementation.